core/src/main/java/org/mule/runtime/core/exception/ErrorTypeLocator.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.exception;
+
+import static java.util.Optional.empty;
+import static org.mule.runtime.core.util.Preconditions.checkArgument;
+import static org.mule.runtime.core.util.Preconditions.checkState;
+
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.commons.collections.map.HashedMap;
+
+import org.mule.runtime.api.message.ErrorType;
+import org.mule.runtime.core.config.ComponentIdentifier;
+
+/**
+ * Locator for error types.
+ *
+ * The locator is responsible for getting the error type of any exception.
+ *
+ * An exception can be a general exception or an exception thrown by a configured component.
+ *
+ * There's a default mapping that will be used when there's no specific mapping between a component and the exception type thrown.
+ *
+ * To create an {@code ErrorTypeLocator} you must use the {@code Builder}. An instance of the builder can be created using the
+ * static method {@code #builder}.
+ * 
+ * @since 4.0
+ */
+public class ErrorTypeLocator {
+
+  private final ErrorTypeRepository errorTypeRepository;
+  private ExceptionMapper defaultExceptionMapper;
+  private Map<ComponentIdentifier, ExceptionMapper> componentExceptionMappers;
+
+  private ErrorTypeLocator(ExceptionMapper defaultExceptionMapper,
+                           Map<ComponentIdentifier, ExceptionMapper> componentExceptionMappers,
+                           ErrorTypeRepository errorTypeRepository) {
+    this.defaultExceptionMapper = defaultExceptionMapper;
+    this.componentExceptionMappers = componentExceptionMappers;
+    this.errorTypeRepository = errorTypeRepository;
+  }
+
+  /**
+   * Gets the {@code ErrorType} instance for ANY error type.
+   * 
+   * @return the ANY error type
+   */
+  public ErrorType getAnyErrorType() {
+    return errorTypeRepository.getAnyErrorType();
+  }
+
+  /**
+   * Finds the {@code ErrorType} related to the provided {@code exception} based on the general mapping rules of the runtime.
+   * 
+   * @param exception the exception related to the error type
+   * @return the error type related to the exception. If there's no mapping then the error type related to UNKNOWN will be
+   *         returned.
+   */
+  public ErrorType lookupErrorType(Exception exception) {
+    return defaultExceptionMapper.resolveErrorType(exception).get();
+  }
+
+  /**
+   * Finds the {@code ErrorType} related to a component defined by the {@link ComponentIdentifier} based on the exception thrown
+   * by the component and the mappings configured in the {@code ErrorTypeLocator}.
+   * 
+   * If no mapping is available then the {@link #getAnyErrorType()} rules applies.
```

Comment:
The above says UNKNOWN but this one ANY in case of no mapping, which one is correct?